duet/cra.ml

```diff
@@ -751,21 +787,114 @@ let decorate_transition_system predicates ts entry =
       let invariant =
         abstract_domain.formula_of (abstract_domain.exists (member live) (inv v))
       in
+      logf "loop header location: %d" v;
       logf "Found invariant at %d:@;%a" v (Syntax.Formula.pp srk) invariant;
+      logf "New ID: %d" fresh_id; 
       WG.split_vertex ts v (Weight (K.assume invariant)) fresh_id)
     ts
 
-let make_transition_system rg =
+module VSet = BatSet.Make(V)
+module ISet = BatSet.Make(Int)
+
+let create_gas_variable () = 
+  let mk_int k = Ctx.mk_real (QQ.of_int k) in 
+  let gas_var = Var.mk (Varinfo.mk_global "__duet_gas" (Concrete (Int 8))) in 
+  let gas_var_sym = Syntax.mk_symbol srk ~name:"__duet_gas" `TyInt in  
+  let gas_var_term = Syntax.mk_const srk gas_var_sym in 
+  Hashtbl.add V.sym_to_var gas_var_sym (VVal gas_var);
+  ValueHT.add V.var_to_sym (VVal gas_var) gas_var_sym;
+  let gasexpr = (Syntax.mk_lt srk (mk_int 0) gas_var_term) in 
+  let gasweight =
+    let assume_positive = K.assume gasexpr in
+    let decr_by_one = Syntax.mk_sub srk gas_var_term (mk_int 1) |> K.assign (VVal gas_var) in
+      K.mul assume_positive decr_by_one in 
+  let initial_gas = K.havoc [ VVal gas_var ] in  
+  (gasweight, gasexpr, initial_gas) 
+
+let new_vtx () = (Def.mk (Assume Bexpr.ktrue)).did 
+
+let instrument_with_gas (ts: TSG.t) gasexpr : TSG.t = 
+  let modify_pre ts u =
```

Comment:
There's a convenient function in ``WeightedGraph`` that helps with this kind of instrumentation---``split_vertex wg u weight v`` takes all of the outgoing edges of u and makes them outgoing edges of v, and adds an edge from u to v with the given weight.